### DISCUSSION TITLE
State of async mutations

### DISCUSSION THREAD

**Discussion by cfneuhaus** [pmndrs/jotai#2177] [Q&A] [answered]
Hi,

I am new to Jotai and I am trying to implement a simple login procedure. As such, I have created this atom:

```
interface LoginStateType {
    username: string,
    jwtToken: string,
}

const loginStateAtom = atomWithStorage<LoginStateType | null>('loginStateAtom', null)
```

It is supposed to store the username + jwt token in local storage so that the login is remembered. Hope that make sense so far.
What I am confused about it how to wrap the async "login" function. I read about async write only atoms and came up with this:

```
export const loginAtom = atom(
    null,
    async (get, set, update: any) => {
        const res = await fetch(.... omitted here... );
        const ret = await res.json()
        set(loginStateAtom, { username: update.ldapUsername, jwtToken: ret.jwtToken });
    }
);
```

That appears to work and in my React component, I can do

```
const [, do_login] = useAtom(loginAtom);

... 
do_login(username, password)
```

Now obviously, I also want to monitor whether the login is in progress or when there is an error. Now this is the part where I am getting confused. What is the best way to achieve this? I can obviously manually create atoms that keep track of this, but this does not feel like the way to go.
How could I use Suspense and ErrorBoundary for this? Is that a good idea in this case?
And alternatively, how could I get a state like loading, error, done for this login process?

I also checked out jotai-tanstack-query, where atomsWithMutation appears to support this. In that case, I could certainly use the mutationFn to do the login, but in that case, I am unsure how I would update my loginStateAtom, which is a different type than what my fetch returns...

Can anyone shed some light on this?

Thank you
   Frank

**Comment by dai-shi** [comment, marked as answer]
> I can obviously manually create atoms that keep track of this

That's a good way. You should be able to keep track of "loading, error, done" states.
I'd suggest to try this first.

> How could I use Suspense and ErrorBoundary for this?

Now, it's a bit tricky.
Check out this post: https://blog.axlight.com/posts/how-to-use-jotai-and-use-transition-for-mutation/
Hm, but I don't know if it works with `atomWithStorage`.
If it doesn't, I'd probably not use `atomWithStorage` and store into localStorage in the write-only atom.

**Reply by cfneuhaus** [reply]
Sorry for the late reply. Thank you, you solved my issue :-)
